subworkflows/local/align_cellrangermulti.nf

```diff
@@ -0,0 +1,244 @@
+            ch_grouped_fastq.gex,
+            ch_grouped_fastq.vdj,
+            ch_grouped_fastq.ab,
+            ch_grouped_fastq.beam,
+            ch_grouped_fastq.cmo,
+            ch_grouped_fastq.crispr,
+            ch_cellranger_gex_index,
+            ch_gex_frna_probeset,
+            ch_gex_target_panel,
+            ch_cellranger_vdj_index,
+            ch_vdj_primer_index,
+            ch_fb_reference,
+            ch_beam_antigen_panel_csv,
+            ch_beam_control_panel_csv,
+            ch_gex_cmo_set,
+            ch_cmo_barcode_csv,
+            ch_gex_barcodes,
+            ch_frna_sample_csv,
+            params.skip_cellranger_renaming
+        )
+        ch_versions = ch_versions.mix(CELLRANGER_MULTI.out.versions)
+
+        //
+        // Cellranger multi splits the results from each sample. So, a module execution will have: (1) a raw counts dir for all;
+        // (2) a filtered counts dir PER sample.
+        //
+        // Thus, cellranger multi outputs data from all identified samples in a single channel, which will cause file collision.
+        //
+        // For the conversion, we should convert the resulting files of each sample, thus, now, we must parse the names
+        // of the filtered 'per_sample_outs' of cellranger/multi and then, split the channels as such.
+        //
+        CELLRANGER_MULTI.out.outs
+        .map{ meta, mtx_files ->
+            def desired_files = []
+            mtx_files.each{
+                if ( it.toString().contains("per_sample_outs") ) {
+                    def demuxed_sample_name = it.toString().split('per_sample_outs/')[1].split('/')[0]
+                    def meta_clone    = meta.clone()
+                    meta_clone.bkp_id = meta_clone.id
+                    meta_clone.id     = demuxed_sample_name
+                    desired_files.add( [ meta_clone, it ] )
+                }
+            }
+            desired_files
+        }
+        .flatten()
+        .buffer( size: 2 )   // gets: [ meta_clone, single_file ]
+        .groupTuple( by: 0 ) // gets: [ meta_clone, all_files   ]
+        .set { ch_parsed_per_sample_matrices }
+
+        //
+        // Split channels of raw and filtered to avoid file collision problems when loading the inputs in conversion modules.
+        //
+        ch_matrices_raw =
+        CELLRANGER_MULTI.out.outs.map { meta, mtx_files -> // Use the unparsed matrix channel because raw results will be outside the 'per_sample_outs'
+            def desired_files = []
+            mtx_files.each{
+                if (
+                        it.toString().contains("raw_feature_bc_matrix") &&
+                        // frna analysis also produces raw data per sample. For standardization, let's keep raw conversions always from the 'unmultiplexed' results.
+                        !(it.toString().contains("sample_raw_feature_bc_matrix")) // TODO: Define if this is enough or if one still wants the raw conversion per samples when frna.
```

Comment:
I think we want them, as we'd need them 
* as input for emptyDrops
* as input for cellbender (downstream of this pipeline)

The demultiplexed samples are not to be treated any different than samplese generated wit hthe sandard protocol.

Reply:
Hi @grst ,
It seems that multiqc still does not have a module for `cellranger multi` results.
https://github.com/MultiQC/MultiQC/issues/2103

Reply:
Hi @grst ,
I have updated the `cellranger multi` subworkflow, so, as you described, I treat the matrices regardless.

I first separate them as raw / filtered. Then, on which, I check if the are from the `per_out_samples` which means, in the cellranger paths, the results specific for the sample named after demultiplexing. On these ones, I modify the `meta.id` so that, during conversion, they receive the correct naming.

Then we have it for both multiplexed and demultiplexed results:

```console
/workspace/scrnaseq/testing (247-support-for-10x-ffpe-scrna) $ tree -L 2 results/cellrangermulti/mtx_conversions/
results/cellrangermulti/mtx_conversions/
├── 4PLEX_HUMAN
│   ├── 4PLEX_HUMAN_raw_matrix.h5ad
│   └── 4PLEX_HUMAN_raw_matrix.rds
├── Colorectal_BC3
│   ├── Colorectal_BC3_filtered_matrix.h5ad
│   ├── Colorectal_BC3_filtered_matrix.rds
│   ├── Colorectal_BC3_raw_matrix.h5ad
│   └── Colorectal_BC3_raw_matrix.rds
├── combined_filtered_matrix.h5ad
├── combined_raw_matrix.h5ad
├── Liver_BC1
│   ├── Liver_BC1_filtered_matrix.h5ad
│   ├── Liver_BC1_filtered_matrix.rds
│   ├── Liver_BC1_raw_matrix.h5ad
│   └── Liver_BC1_raw_matrix.rds
├── Ovarian_BC2
│   ├── Ovarian_BC2_filtered_matrix.h5ad
│   ├── Ovarian_BC2_filtered_matrix.rds
│   ├── Ovarian_BC2_raw_matrix.h5ad
│   └── Ovarian_BC2_raw_matrix.rds
├── Pancreas_BC4
│   ├── Pancreas_BC4_filtered_matrix.h5ad
│   ├── Pancreas_BC4_filtered_matrix.rds
│   ├── Pancreas_BC4_raw_matrix.h5ad
│   └── Pancreas_BC4_raw_matrix.rds
├── PBMC_10K
│   ├── PBMC_10K_filtered_matrix.h5ad
│   ├── PBMC_10K_filtered_matrix.rds
│   ├── PBMC_10K_raw_matrix.h5ad
│   └── PBMC_10K_raw_matrix.rds
├── PBMC_10K_CMO
│   ├── PBMC_10K_CMO_raw_matrix.h5ad
│   └── PBMC_10K_CMO_raw_matrix.rds
├── PBMC_10K_CMO_PBMCs_human_1
│   ├── PBMC_10K_CMO_PBMCs_human_1_filtered_matrix.h5ad
│   └── PBMC_10K_CMO_PBMCs_human_1_filtered_matrix.rds
├── PBMC_10K_CMO_PBMCs_human_2
│   ├── PBMC_10K_CMO_PBMCs_human_2_filtered_matrix.h5ad
│   └── PBMC_10K_CMO_PBMCs_human_2_filtered_matrix.rds
├── PBMC_10K_CMV
│   ├── PBMC_10K_CMV_filtered_matrix.h5ad
│   ├── PBMC_10K_CMV_filtered_matrix.rds
│   ├── PBMC_10K_CMV_raw_matrix.h5ad
│   └── PBMC_10K_CMV_raw_matrix.rds
└── versions.yml
```

Reply:
perfect!

Reply:
However, I am still getting an empty multiqc.
Which is expected for the `cellranger multi` results which are not supported. But, is not expected for the `fastqc` results.
I will try to figure it out tomorrow.

From the comments, this is the last thing missing.

Reply:
> should I also execute emptydrops filter on the cellranger multi results?

I don't see why not to, but if it doesn't work straight away, I wouldn't invest more time on this for now. 

> would need help in providing descriptions for the input files and documentation, as I am not very familiar with all data files.

Can you get started with the documentation, I'm happy to give it another iteration then.

Reply:
> However, I am still getting an empty multiqc. Which is expected for the `cellranger multi` results which are not supported. But, is not expected for the `fastqc` results. I will try to figure it out tomorrow.
> 
> From the comments, this is the last thing missing.

Report now looks fine. I had forgotten to add the out channels with a .mix() operator, so I was overwriting instead of adding.

Now it is fixed, and reports have FASTQC results, but not cellranger multi results as it is not supported.

Reply:
Since `emptydrops` does not work on small dataset, I am running it to check if it works with a [full 10X dataset](https://github.com/nf-core/test-datasets/tree/modules/data/genomics/homo_sapiens/10xgenomics/cellranger/10k_pbmc_cmo).

Then we would have from [this](https://github.com/nf-core/scrnaseq/pull/276#issuecomment-2084997587), the final showdown:

- [ ] emptydrops answer
- [ ] documentation started

Reply:
Hi @grst ,
Using a full dataset, I was able to successfully include cellrangermulti to pass to the custom emptydrops. It worked out of the box. 
Just as the others, it works receiving the "raw" matrices so that it can perform the custom filtering as it was described in the PR we added the feature.

```console
tree -L 2 FULL_DATASET_CMULTI_RESULTS/cellrangermulti/mtx_conversions/
FULL_DATASET_CMULTI_RESULTS/cellrangermulti/mtx_conversions/
├── combined_custom_emptydrops_filter_matrix.h5ad
├── combined_filtered_matrix.h5ad
├── combined_raw_matrix.h5ad
├── PBMC_10K_CMO
│   ├── PBMC_10K_CMO_custom_emptydrops_filter_matrix.h5ad
│   ├── PBMC_10K_CMO_custom_emptydrops_filter_matrix.rds
│   ├── PBMC_10K_CMO_raw_matrix.h5ad
│   └── PBMC_10K_CMO_raw_matrix.rds
├── PBMC_10K_CMO_PBMCs_human_1
│   ├── PBMC_10K_CMO_PBMCs_human_1_filtered_matrix.h5ad
│   └── PBMC_10K_CMO_PBMCs_human_1_filtered_matrix.rds
├── PBMC_10K_CMO_PBMCs_human_2
│   ├── PBMC_10K_CMO_PBMCs_human_2_filtered_matrix.h5ad
│   └── PBMC_10K_CMO_PBMCs_human_2_filtered_matrix.rds
└── versions.yml

3 directories, 12 files
```

Now, the only thing missing is the documentation which I will start and flag you on it.

Reply:
Hi @grst ,
I have started the documentation.
Did the part I felt comfortable with.

Added TODOs in the parts or sections that need a better explanation on what is expected and what they are, and I think is best if someone "closer" to the data writes them.

Files: `docs/{usage,output}.md`.

Cheers.